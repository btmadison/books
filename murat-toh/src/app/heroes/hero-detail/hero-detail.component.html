<div *ngIf="hero">
  <h2>{{hero.name | uppercase}} Details</h2>
  <div><span>id: </span>{{hero.id}}</div>
  <div>
    <label>name:
      <input [(ngModel)]="hero.name" placeholder="name" />
    </label>
  </div>
  <p>
    <button (click)="save()">save</button>
    <button (click)="goBack()">go back</button>
    <button (click)="gotoHeroes(hero)">go to heroes</button>
  </p>
</div>
