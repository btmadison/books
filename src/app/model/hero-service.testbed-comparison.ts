import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';

import { TestBed } from '@angular/core/testing';
import { HttpResponse } from '@angular/common/http';

import { Hero } from './hero';
import { HeroService } from './hero.service';


// mocking external core services like Http
// to setup create an httpMock with the help of HttpTestingController from HttpClientTestingModule (3.1)
// prep hardcoded data, initiate the client request and setup the assertion that will happen, match the url w/ httpMock.expectOne, using httpMock.expectOne, (3.2)
// use .flush to send the data to the client and .error to emulate error (3.3)
// in error testing, define the error response and bypass the success case (3.4)
// cover the Multiple Request Case, and check the request length (3.5)
// testing PUTs: test the method type and request body that is going out from the client, Use req.event(new HttpResponse({ .. })) to respond' (3.6)

describe('Testing Http with TestBed', () => {
  let heroService: HeroService;
  let httpMock: HttpTestingController;

  // for the hardcoded data that is shared between the tests
  let expectedHeroes: Hero[];
  let updatedHero: Hero;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        HttpClientTestingModule // (3.1.1) HttpClientTestingModule in imports
      ],
      providers: [HeroService]
    });

    // (3.1.2) inject the service under test (just like 2.2 )
    heroService = TestBed.inject(HeroService);
    // (3.1.3) mock http using  HttpTestingController
    httpMock = TestBed.inject(HttpTestingController);
    // (service injection similar to other external services, the key difference is the module being imported)
    // (HttpTestingController doesnâ€™t make an HTTP request but allows you to emulate it using hardcoded data.)
  });

  // verify() is used to assert that there are no outstanding http requests
  afterEach(() => httpMock.verify());

  describe('GET requests' , () => {

    beforeEach(() => {
      // (3.2.1) prepare hardcoded data that will be used as the response
      expectedHeroes = [
        { id: 1, name: 'A' },
        { id: 2, name: 'B' },
      ] as Hero[];
    });

    it('(3.2.2) initiate the client request and setup the assertion that will happen,(3.2.3) match the url w/ httpMock.expectOne, (3.3) use .flush to respond to the client ', () => {

      // (3.2.2) initiate the client request, and setup the assertion that will happen once the observable is fulfilled
      heroService.getHeroes().subscribe(
        heroes => expect(heroes).toEqual(expectedHeroes),
        fail // the error case of the observable
      );

      // (3.2.3) expect that a single request has been made which matches the given URL, using httpMock.expectOne
      const req = httpMock.expectOne(heroService.heroesUrl);
      // (3.3) use .flush to respond to the client
      req.flush(expectedHeroes);

      expect(req.request.method).toEqual('GET'); // extra
    });

    it('cover the Empty Response Case', () => {
      heroService.getHeroes().subscribe(
        heroes => expect(heroes.length).toEqual(0),
        fail
      );

      const req = httpMock.expectOne(heroService.heroesUrl);

      req.flush([]); // we control the response here
    });

    it('(3.4) in error testing, define the error response and bypass the success', () => {
      const msg = 'Deliberate 404';

      heroService.getHeroes().subscribe(
        heroes => fail('expected to fail'),
        error => expect(error.message).toContain(msg) // 3.4.0 in error testing, define the error response and bypass the success case
      );

      const req = httpMock.expectOne(heroService.heroesUrl);

      // 3.4.1 flush with the optional object: status and message
      req.flush(msg, {status: 404, statusText: 'Not Found'});
    });

    it('(3.5) cover the Multiple Request Case, and check the request length', () => {
      heroService.getHeroes().subscribe(); // KEY multiple subscribes
      heroService.getHeroes().subscribe();
      heroService.getHeroes().subscribe(
        heroes => expect(heroes).toEqual(expectedHeroes),
        fail
      );

      const requests = httpMock.match(heroService.heroesUrl);

      // Respond to each request with different mock hero results
      requests[0].flush([]);
      requests[1].flush([{ id: 1, name: 'bob' }]);
      requests[2].flush(expectedHeroes);

      expect(requests.length).toEqual(3); // KEY check request length
    });
  });

  describe('PUT requests', () => {

    beforeEach(() => {
      updatedHero = { id: 1, name: 'A' } as Hero;
    });

    it('(3.6) testing PUTs: test the method type and request body that is going out from the client, Use req.event(new HttpResponse({ .. })) to respond', () => {

      heroService.updateHero(updatedHero).subscribe(
        data => expect(data).toEqual(updatedHero),
        fail
      );

      const req = httpMock.expectOne(heroService.heroesUrl);
      // (3.6.0) KEY with testing PUTs: test the method type and request body that is going out from the client
      expect(req.request.method).toEqual('PUT');
      expect(req.request.body).toEqual(updatedHero);

      // (3.6.1) KEY with PUT: use req.event(new HttpResponse({})) instead of req.flush()
      req.event(
        new HttpResponse({ status: 200, statusText: 'OK', body: updatedHero })
      );
    });

    it('cover the Error Case, same as the GET scenario in (3.4) ', () => {
      const msg = 'Deliberate 404';

      heroService.updateHero(updatedHero).subscribe(
        heroes => fail('expected to fail'),
        error => expect(error.message).toContain(msg) // KEY (same as GET scenario IN 3.4) in error testing, define the error response and bypass the success
      );

      const req = httpMock.expectOne(heroService.heroesUrl);

      // KEY: (same as GET scenario) flush with the optional object: status and message
      req.flush(msg, { status: 404, statusText: 'Not Found' });
    });
  });
});
