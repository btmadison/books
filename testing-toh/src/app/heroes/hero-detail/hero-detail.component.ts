import { Component, OnInit, Input } from '@angular/core';
import { Router, ActivatedRoute, ParamMap } from '@angular/router';
import { Hero } from '../../_core/models/hero';
import { HeroApiService } from '../../_core/hero-api.service';
import { Location } from '@angular/common';

@Component({
  selector: 'app-hero-detail',
  templateUrl: './hero-detail.component.html',
  styleUrls: ['./hero-detail.component.css'],
})
export class HeroDetailComponent implements OnInit {
  hero: Hero;

  constructor(
    private heroService: HeroApiService,
    private router: Router,
    private route: ActivatedRoute,
    private location: Location
  ) {}

  ngOnInit() {
    this.getHero();
  }

  getHero(): void {
    const id = +this.route.snapshot.paramMap.get('id');
    this.heroService.getHero(id)
      .subscribe(hero => this.hero = hero);
  }

  save(): void {
    this.heroService.updateHero(this.hero)
      .subscribe(() => this.goBack());
  }

  gotoHeroes(hero: Hero) {
    const heroId = hero ? hero.id : null;
    this.router.navigate(['/superheroes', { id: heroId, foo: 'foo' }]);
  }

  goBack(): void {
    this.location.back();
  }
}
