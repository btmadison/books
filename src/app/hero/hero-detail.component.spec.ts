import { Router, ActivatedRoute } from '@angular/router';
import { Spectator, createComponentFactory, mockProvider, createRoutingFactory, byText, byTextContent } from '@ngneat/spectator/jest';

import { Hero } from '../model/hero';
import { HeroService } from '../model/hero.service';
import { HeroDetailComponent } from './hero-detail.component';
import { HeroDetailService } from './hero-detail.service';
import { HeroModule } from './hero.module';
import { HeroRoutingModule } from './hero-routing.module';
import { By } from '@angular/platform-browser';


// [4] testing components that use routing // https://github.com/ngneat/spectator#testing-with-routing
// setup the component with routing; use createRoutingFactory to auto-mock Router and ActivatedRoute (4.1)
// use convenience methods to trigger a navigation, if needed (4.2) https://github.com/ngneat/spectator#triggering-a-navigation


describe('module test', () => {
  let component: HeroDetailComponent;
  let spectator: Spectator<HeroDetailComponent>;

  // (4.1) setup the component with routing. Use createRoutingFactory to auto-stub Router and ActivatedRoute
  const createComponent = createRoutingFactory({ // compared to createComponentFactory, Route and ActivatedRoute are auto-mocked
    component: HeroDetailComponent,
    mocks: [HeroDetailService, HeroService],
    detectChanges: false,
    stubsEnabled: false, // (true by default) when stubsEnabled option is false, you can pass a real routing configuration and setup an integration test using the RouterTestingModule from Angular
    params: { id: 41 }, // initial params to use in ActivatedRoute stub
    data: { name: 'Dr Nice' } // initial data to use in ActivatedRoute stub
  });

  beforeEach(() => {
    spectator = createComponent();    // (4.1) setup the component.. TestBed.configureTestingModule({..})
    component = spectator.component;  // (1.2) access the TS...     comp = fixture.debugElement.componentInstance
  });

  it('sanity', () => {
    expect(component).toBeTruthy();
  });

  /* DOM testing library-like approach
    spectator.query(byPlaceholder('Please enter your email address'));
    spectator.query(byValue('By value'));
    spectator.query(byTitle('By title'));
    spectator.query(byAltText('By alt text'));
    spectator.query(byLabel('By label'));
    spectator.query(byText('By text'));
    spectator.query(byText('By text', {selector: '#some .selector'}));
    spectator.query(byTextContent('By text content', {selector: '#some .selector'}));
  */

  it('query byText', () => {
    expect(spectator.query(byText('Dr Nice'))).toBeDefined();
    spectator.query(byTextContent('Dr Nice', {selector: '.qa-hero-name'}));
  });

  it('should navigate when click cancel', async () => {
    spectator.detectChanges();
    spectator.query('.qa-cancel'); // exists
    spectator.query(byText('Cancel')); // exists;

    // spectator.click('.qa-cancel'); // can't click
    // spectator.click(byText('Cancel')); // TODO: click it , but doesn't work
    // expect(page.navigateSpy.calls.any()).toBe(true, 'router.navigate called');
  });

  // convert to spectator

  // it('should save when click save but not navigate immediately', () => {
  //   // Get service injected into component and spy on its`saveHero` method.
  //   // It delegates to fake `HeroService.updateHero` which delivers a safe test result.
  //   const hds = fixture.debugElement.injector.get(HeroDetailService);
  //   const saveSpy = spyOn(hds, 'saveHero').and.callThrough();

  //   click(page.saveBtn);
  //   expect(saveSpy.calls.any()).toBe(true, 'HeroDetailService.save called');
  //   expect(page.navigateSpy.calls.any()).toBe(false, 'router.navigate not called');
  // });

  // it('should navigate when click save and save resolves', fakeAsync(() => {
  //      click(page.saveBtn);
  //      tick();  // wait for async save to complete
  //      expect(page.navigateSpy.calls.any()).toBe(true, 'router.navigate called');
  //    }));

  // it('should convert hero name to Title Case', () => {
  //   // get the name's input and display elements from the DOM
  //   const hostElement = fixture.nativeElement;
  //   const nameInput: HTMLInputElement = hostElement.querySelector('input');
  //   const nameDisplay: HTMLElement = hostElement.querySelector('span');

  //   // simulate user entering a new name into the input box
  //   nameInput.value = 'quick BROWN  fOx';

  //   // Dispatch a DOM event so that Angular learns of input value change.
  //   // In older browsers, such as IE, you might need a CustomEvent instead. See
  //   // https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent#Polyfill
  //   nameInput.dispatchEvent(new Event('input'));

  //   // Tell Angular to update the display binding through the title pipe
  //   fixture.detectChanges();

  //   expect(nameDisplay.textContent).toBe('Quick Brown  Fox');
  // });
});

