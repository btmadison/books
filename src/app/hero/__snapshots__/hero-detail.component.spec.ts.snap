// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module test async with await spectator.fixture.whenStable() version: should navigate when clicking cancel 1`] = `
<app-hero-detail
  __ngContext__={[Function LRootView]}
  hero={[Function Object]}
  heroDetailService={[Function HeroDetailService]}
  route={[Function ActivatedRouteStub]}
  router={[Function Object]}
>
  <div>
    <h2>
      <span
        class="qa-hero-name"
      >
        
      </span>
       Details
    </h2>
    <div
      class="qa-hero-id"
    >
      <label>
        id: 
      </label>
      0
    </div>
    <div>
      <label
        for="name"
      >
        name: 
      </label>
      <input
        class="ng-untouched ng-pristine ng-valid"
        id="name"
        placeholder="name"
      />
    </div>
    <button
      class="qa-save"
    >
      Save
    </button>
    <button
      class="qa-cancel"
    >
      Cancel
    </button>
  </div>
</app-hero-detail>
`;

exports[`module test query byText 1`] = `
<app-hero-detail
  __ngContext__={[Function LRootView]}
  hero={[Function Object]}
  heroDetailService={[Function HeroDetailService]}
  route={[Function ActivatedRouteStub]}
  router={[Function Object]}
>
  <div>
    <h2>
      <span
        class="qa-hero-name"
      >
        
      </span>
       Details
    </h2>
    <div
      class="qa-hero-id"
    >
      <label>
        id: 
      </label>
      0
    </div>
    <div>
      <label
        for="name"
      >
        name: 
      </label>
      <input
        class="ng-untouched ng-pristine ng-valid"
        id="name"
        placeholder="name"
      />
    </div>
    <button
      class="qa-save"
    >
      Save
    </button>
    <button
      class="qa-cancel"
    >
      Cancel
    </button>
  </div>
</app-hero-detail>
`;

exports[`module test sanity 1`] = `
<app-hero-detail
  __ngContext__={[Function LRootView]}
  hero={[Function Object]}
  heroDetailService={[Function HeroDetailService]}
  route={[Function ActivatedRouteStub]}
  router={[Function Object]}
>
  <div>
    <h2>
      <span
        class="qa-hero-name"
      >
        
      </span>
       Details
    </h2>
    <div
      class="qa-hero-id"
    >
      <label>
        id: 
      </label>
      0
    </div>
    <div>
      <label
        for="name"
      >
        name: 
      </label>
      <input
        class="ng-untouched ng-pristine ng-valid"
        id="name"
        placeholder="name"
      />
    </div>
    <button
      class="qa-save"
    >
      Save
    </button>
    <button
      class="qa-cancel"
    >
      Cancel
    </button>
  </div>
</app-hero-detail>
`;
